blaze/utils.py

```diff
@@ -157,7 +160,74 @@ def listpack(x):
 
 @dispatch(datetime.datetime)
 def json_dumps(dt):
-    s = dt.isoformat()
-    if not dt.tzname():
-        s += 'Z'
-    return s
+    if dt is pd.NaT:
+        # NaT has an isoformat but it is totally invalid.
+        # This keeps the parsing on the client side simple.
+        s = 'NaT'
+    else:
+        s = dt.isoformat()
+        if not dt.tzname():
+            s += 'Z'
+
+    return {'__!datetime': s}
+
+
+@dispatch(frozenset)
+def json_dumps(ds):
+    return {'__!frozenset': list(ds)}
+
+
+@dispatch(datetime.timedelta)
+def json_dumps(ds):
+    return {'__!timedelta': ds.total_seconds()}
+
+
+def object_hook(obj):
+    """Convert a json object dict back into a python object.
+
+    This looks for our objects that have encoded richer representations with
+    a ``__!{type}`` key.
+
+    Parameters
+    ----------
+    obj : dict
+        The raw json parsed dictionary.
+
+    Returns
+    -------
+    parsed : any
+        The richer form of the object.
+
+    Notes
+    -----
+    The types that this reads can be extended with the ``register`` method.
+    For example:
+
+    >>> class MyList(list):
+    ...     pass
+    >>> @object_hook.register('MyList')
+    ... def _parse_my_list(obj):
+    ...     return MyList(obj)
+
+    Register can also be called as a function like:
+    >>> object_hook.register('frozenset', frozenset)
+    """
+    if len(obj) != 1:
+        return obj
+
+    key, = obj.keys()
+    if not key.startswith('__!'):
+        return obj
+
+    return object_hook._converters[key[3:]](obj[key])
```

Comment:
can you write this as `key[len('__!'):]`? IMO this clarifies what you're doing here